core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -419,6 +419,11 @@ object Endpoint {
    */
   type Module[F[_]] = EndpointModule[F]
 
+  /**
+    * Representation of function Request => F[(Trace, Response)]
+    */
+  type Compiled[F[_]] = Kleisli[F, Request, (Trace, Response)]
```

Comment:
One thing we may consider with regards to `Trace` is to embed it into our `Error` model, such that if `IO` resolves into `io.finch.Error`, it will have a `trace: Trace` member.

Not exactly sure I like this, just throwing out ideas.

Reply:
Customers might throw their own exceptions

Reply:
I mean, we can wrap them. Not sure it's a great way of doing this but it's workable.

Reply:
What about `Request => (Trace, F[Response])`? I know, it's not Kleisli anymore, which is sad but on the other hand it's minimum allocations and a guarantee of having tracing all the time

Reply:
I'd probably prefer us to stay within a "normal" `Kleisli` as it gives us insane composability for free.

We can also re-consider shoving `Trace` onto both `Request.ctx` and `Response.ctx` as we originally discussed.

Reply:
We sort of already doing `handle` within `Compile` and pretty much convert any unhandled exceptions into `io.finch.Error` (please, correct me if I'm wrong). That said, there is nothing else can expect in `Compiled` beyond either successful `Response` or an `Error`. Why not do `Error.trace` than? Seems very reasonable to me to be honest (assuming I'm correct about users exceptions not being propagated there).

Reply:
No, I don't think so. I'm sleepy already but what I see is:

```scala
private val respond400: PartialFunction[Throwable, Output[Nothing]] = {
    case e: io.finch.Error => Output.failure(e, Status.BadRequest)
    case es: io.finch.Errors => Output.failure(es, Status.BadRequest)
  }

  private val respond415: PartialFunction[Throwable, Output[Nothing]] = {
    case e: io.finch.Error if e.getCause eq Decode.UnsupportedMediaTypeException =>
      Output.failure(e, Status.UnsupportedMediaType)
  }
```

We don't catch & wrap user exceptions here

Reply:
Honestly, more I think about it and try it locally, more I doubt the idea of
`Kleisli[F, Request, Response]`.

Having something like `Request => (Trace, F[Response])` solves the case of happy/unhappy `Trace` almost for free and cases with complex filters where you want to have both `Request` and `F[Response]` (i.e. monitoring, time measurements and distributed tracing). While `Kleisli` helps with simple situations in `compose` and `andThen`, it also brings additional complexity when we go outside the usual composition.

So I'd go with our custom `Filter` or similar name. We can boost it with cats typeclasses if it's neccesary

Reply:
Beyond `Trace` is there anything else that bothers you? I'm not convinced we should give up on the Kleisli idea this easy. Both of the approaches of handling taces seem pretty reasonable to me. Unless there is something else concerning about `Kleisli`, I'd probably prefer we keep discussing it.

http4s uses Kleilsi as its middleware and they seem to handle things like this.

Reply:
- Embedding user exceptions into `io.finch.Error` might lead to unexpected behavior for users when their error handlers made on top of pattern matching don't work anymore. The "best" part that they wouldn't even notice any warnings in compile time in that case.
- `F.attempt` is going to hit benchmarks for sure, also turning the type signature into a nice mess.

Could you provide the example of how the Kleisli based filter and their composition would look like? May be I just can't express it nicely

Reply:
I can give it a try based on my very limited exposure that Klesili. 

Auth could be a `flatMap`:

```scala
scala> val in = Kleisli.ask[cats.effect.IO, Request]
in: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Request] = Kleisli(cats.data.KleisliFunctions$$Lambda$5376/280682234@58b00155)

scala> val middleware = Kleisli.pure[cats.effect.IO, Request, Response](Response())
middleware: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Response] = Kleisli(cats.data.KleisliFunctions$$Lambda$5396/1176582514@70aa5f1)

scala> in.flatMap { req => if (req.content.length > 0) middleware else Kleisli.pure[cats.effect.IO, Request, Response](Response(Status.BadRequest)) }
res1: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Response] = Kleisli(cats.data.Kleisli$$$Lambda$5401/1855768135@24fff07f)
```

Something like logging would probably be `andThen`:

```scala
scala> val logging = Kleisli.ask[cats.effect.IO, Request].map { req => println(req); req }
logging: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Request] = Kleisli(cats.data.Kleisli$$Lambda$5403/1811915277@36350e9e)

scala> logging.andThen(middleware)
res2: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Response] = Kleisli(cats.data.Kleisli$$$Lambda$5401/1855768135@648c6ebb)
```

`tapWith` would allow to extend logging on responses too:

```scala
scala> middleware.tapWith { (req, rep) => println(req); println(rep); rep }

res3: cats.data.Kleisli[cats.effect.IO,com.twitter.finagle.http.Request,com.twitter.finagle.http.Response] = Kleisli(cats.data.Kleisli$$Lambda$5492/986060877@2eb34475)
```

Stats can probably be a bimap:

```scala
scala> middleware.bimap(req => ({println("start"); req}), rep => ({println("finish"); rep}))
```

Reply:
So exposing Kleisli isn't necessarily mean limiting options. I actually believe it's the opposite: in addition to the combinators `Kleisli` already expose, you can also drop down to a filter-like structure discussed in the description of this PR that just wraps `Kleisli`. I'm not super sure yet it should be our responsibility to determine a shape of those filter-like structures. In my mind's eye, these are application-level concerns.

I'd be curious to see the benchmark numbers for `IO[Either[Trace -> Throwable, Trace -> Response]`. I'm doubtful the penalty is that bad. The ergonomics, Kleisli provide seems to outweigh the anticipated performance hit, keeping in mind we've been always overserving in performance. I'd be happy to trade some of it for developer's productivity.

Reply:
So I've changed the signature to:
```scala
type Endpoint.Compiled[F] = Kleisli[F, Request, (Trace, Either[Throwable, Response])]
```
and made some measurements. Actually, it's not _that_ bad.

**Before** (master):
```
[info] Benchmark                                                                    Mode  Cnt       Score       Error   Units
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos                                thrpt   10   32358.348 ±  1809.015   ops/s
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.alloc.rate                 thrpt   10    1609.298 ±    90.256  MB/sec
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.count                      thrpt   10      74.000              counts
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.time                       thrpt   10     143.000                  ms
[info] JsonBootstrapBenchmark.foos                                                 thrpt   10   22776.430 ±  1863.079   ops/s
[info] JsonBootstrapBenchmark.foos:·gc.alloc.rate                                  thrpt   10    1503.841 ±   123.266  MB/sec
[info] JsonBootstrapBenchmark.foos:·gc.count                                       thrpt   10      70.000              counts
[info] JsonBootstrapBenchmark.foos:·gc.time                                        thrpt   10     117.000                  ms
[info] TextBootstrapBenchmark.foos                                                 thrpt   10   64812.622 ±  6849.459   ops/s
[info] TextBootstrapBenchmark.foos:·gc.alloc.rate                                  thrpt   10    1972.065 ±   209.078  MB/sec
[info] TextBootstrapBenchmark.foos:·gc.count                                       thrpt   10      92.000              counts
[info] TextBootstrapBenchmark.foos:·gc.time                                        thrpt   10     159.000                  ms
```
**After** (this branch):
```
[info] Benchmark                                                                    Mode  Cnt       Score       Error   Units
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos                                thrpt   10   31949.906 ±  4026.098   ops/s
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.alloc.rate                 thrpt   10    1577.434 ±   213.755  MB/sec
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.count                      thrpt   10      72.000              counts
[info] JsonAndTextNegotiatedBootstrapBenchmark.foos:·gc.time                       thrpt   10     147.000                  ms
[info] JsonBootstrapBenchmark.foos                                                 thrpt   10   21939.217 ±  1327.161   ops/s
[info] JsonBootstrapBenchmark.foos:·gc.alloc.rate                                  thrpt   10    1452.948 ±    88.720  MB/sec
[info] JsonBootstrapBenchmark.foos:·gc.count                                       thrpt   10      67.000              counts
[info] JsonBootstrapBenchmark.foos:·gc.time                                        thrpt   10     120.000                  ms
[info] TextBootstrapBenchmark.foos                                                 thrpt   10   65395.311 ±  3172.109   ops/s
[info] TextBootstrapBenchmark.foos:·gc.alloc.rate                                  thrpt   10    2001.061 ±    97.205  MB/sec
[info] TextBootstrapBenchmark.foos:·gc.count                                       thrpt   10      92.000              counts
[info] TextBootstrapBenchmark.foos:·gc.time                                        thrpt   10     154.000                  ms
```